src/algokit_utils/_ensure_funded.py

```diff
@@ -49,52 +45,100 @@ class EnsureBalanceParameters:
     """(optional)The maximum fee that you are happy to pay (default: unbounded) -
     if this is set it's possible the transaction could get rejected during network congestion"""
 
+    use_dispenser_api: bool = False
+    """If True, will use the dispenser API to fund the account if it's on TestNet"""
+
+
+def _get_address_to_fund(parameters: EnsureBalanceParameters) -> str:
+    if isinstance(parameters.account_to_fund, str):
+        return parameters.account_to_fund
+    else:
+        return str(address_from_private_key(parameters.account_to_fund.private_key))  # type: ignore[no-untyped-call]
+
+
+def _get_account_info(client: AlgodClient, address_to_fund: str) -> dict:
+    account_info = client.account_info(address_to_fund)
+    assert isinstance(account_info, dict)
+    return account_info
+
+
+def _calculate_fund_amount(
+    parameters: EnsureBalanceParameters, current_spending_balance_micro_algos: int
+) -> int | None:
+    if parameters.min_spending_balance_micro_algos > current_spending_balance_micro_algos:
+        min_fund_amount_micro_algos = parameters.min_spending_balance_micro_algos - current_spending_balance_micro_algos
+        return max(min_fund_amount_micro_algos, parameters.min_funding_increment_micro_algos)
+    else:
+        return None
+
+
+def _fund_using_dispenser_api(address_to_fund: str, fund_amount_micro_algos: int) -> str | None:
+    ci_access_token = os.environ.get(DISPENSER_ACCESS_TOKEN_KEY)
+
+    if not ci_access_token:
+        raise Exception(
+            f"Can't use dispenser API to fund account {address_to_fund} "
+            "via AlgoKit TestNet Dispenser API because "
+            f"environment variable {DISPENSER_ACCESS_TOKEN_KEY} is not set"
+        )
+
+    return process_dispenser_fund_request(
+        address=address_to_fund,
+        amount=fund_amount_micro_algos,
+        asset_id=DispenserAssetName.ALGO,
+        auth_token=ci_access_token,
+    )
+
+
+def _fund_using_transfer(
+    client: AlgodClient, parameters: EnsureBalanceParameters, address_to_fund: str, fund_amount_micro_algos: int
+) -> PaymentTxn:
+    funding_source = parameters.funding_source or get_dispenser_account(client)
+    return transfer(
+        client,
+        TransferParameters(
+            from_account=funding_source,
+            to_address=address_to_fund,
+            micro_algos=fund_amount_micro_algos,
+            note=parameters.note or "Funding account to meet minimum requirement",
+            suggested_params=parameters.suggested_params,
+            max_fee_micro_algos=parameters.max_fee_micro_algos,
+            fee_micro_algos=parameters.fee_micro_algos,
+        ),
+    )
+
 
 def ensure_funded(
-    client: "AlgodClient",
+    client: AlgodClient,
     parameters: EnsureBalanceParameters,
-) -> None | PaymentTxn:
+) -> PaymentTxn | str | None:
```

Comment:
This return type feels a little strange. Ideally I think we'd keep the same return type.
Additionally `str` in this context is a bit vague and doesn't inform the user that this is a transaction id.

Reply:
@neilcampbell @robdmoore brought the question up on ext channel, doesn't seem like i can reassemble the same txn in python since the txn object on dispenser is encoded and assembled via js sdk.

Reply:
@robdmoore @neilcampbell i haven't found any example of ensureFunded on public github that relies on the output + im taggin latest commit as breaking change so i think a new return type with txn and amount is a best option to go